src/es6/SirenUtil.js

```diff
@@ -0,0 +1,198 @@
+import 'd2l-fetch/d2l-fetch.js';
+import 'd2l-polymer-siren-behaviors/store/entity-store.js';
+import 'd2l-polymer-siren-behaviors/store/action-queue.js';
+import SirenParse from 'siren-parser';
+
+const _getSirenFields = function(action) {
+	const url = new URL(action.href, window.location.origin);
+	const fields = [];
+	if (action.method === 'GET' || action.method === 'HEAD') {
+		for (const param in url.searchParams.entries()) {
+			fields.push({ name: param[0], value: param[1] });
+		}
+		// Disable URLSearchParams until they are fully supported (i.e. Edge)
+		/*
+		} else if (window.URLSearchParams && action.type === 'application/x-www-form-urlencoded') {
+			fields = new URLSearchParams();
+		*/
+	}
+
+	if (action.fields && action.fields.forEach) {
+		action.fields.forEach(function(field) {
+			if (field.value === undefined) {
+				return;
+			}
+			// if the field is specified multiple times, assume it is intentional
+			fields.push({ name: field.name, value: field.value });
+		});
+	}
+	return fields;
+};
+
+const _getEntityUrl = function(action, fields) {
+	if (!action) {
+		return null;
+	}
+
+	let url = new URL(action.href, window.location.origin);
+
+	fields = fields || _getSirenFields(action);
+	if (action.method === 'GET' || action.method === 'HEAD') {
+		const params = _createURLSearchParams(fields);
+		url = new URL(url.pathname + '?' + params.toString(), url.origin);
+	}
+
+	return url;
+};
+
+const _createURLSearchParams = function(fields) {
+	const sequence = [];
+	for (let i = 0; i < fields.length; i++) {
+		const field = fields[i];
+		sequence.push([field.name, field.value]);
+	}
+	return new URLSearchParams(sequence);
+};
+
+const _createFormData = function(fields) {
+	const formData = new FormData();
+	for (let i = 0; i < fields.length; i++) {
+		formData.append(fields[i].name, fields[i].value);
+	}
+	return formData;
+};
+
+const _appendHiddenFields = function(action, fields) {
+	if (action.fields && action.fields.forEach) {
+		action.fields.forEach(function(field) {
+			if (field.type === 'hidden' && field.value !== undefined) {
+				fields.push({ name: field.name, value: field.value });
+			}
+		});
+	}
+	return fields;
+};
+
+const _fetch = function(href, opts) {
+	const methodTypes = ['POST', 'PUT', 'DELETE', 'PATCH'];
+	const sendSaveEvent = methodTypes.indexOf(opts.method) !== -1;
+	const self = this; // eslint-disable-line no-invalid-this
+
+	if (sendSaveEvent) {
+		self.dispatchEvent(new CustomEvent('d2l-siren-entity-save-start', {
+			bubbles: true,
+			composed: true
+		}));
+	}
+	return window.d2lfetch.fetch(href, opts)
+		.then(function(resp) {
+			if (sendSaveEvent && resp.ok) {
+				self.dispatchEvent(new CustomEvent('d2l-siren-entity-save-end', {
```

Comment:
Alternatively, I wonder if you really care about the element that initiated the event or that the entity id has saved it.

Like do i care that course tile x is done saving? or do i care that http://path.to.course.tile.thing is done saving?

Reply:
In our current usage, it's the element as that's the way the listeners are set up - but that could be changed. In reality, we don't actually care where these events are coming from anyway, since there's just the one global-esque autosave component on the page listening for these events. That said, `dispatchEvent` is from `HTMLElement`, so either this utility class needs to extend `LitElement`, or I'll need to figure out another way 🤔